### DISCUSSION TITLE
Is bpftrace binary compatible with all distributions

### DISCUSSION THREAD

**Discussion by Engg-blogger** [bpftrace/bpftrace#2870] [General]
Hi Team - I am planning to use bpftrace binary across multiple distributions. Is the same binary compatible across various distributions?

**Comment by Engg-blogger** [comment]
The reason why I am asking is because i built bpftrace binary on ubuntu 20.04 and then try to run that binary on ubuntu 22.04. But i received an error saying that _bpftrace: error accessing the shared library libbfd-2.34_. When i looked for libbfd, i saw that ubuntu 22 had libbfd-2.38 installed on it. Could you clarify on this?

**Reply by danobi** [reply]
It sounds like you're building bpftrace as a dynamically linked binary. Dynamically linked binaries are not usually portable across major versions of a distro (eg u20 -> u22). This is not special to bpftrace - nearly all linux binaries work like this.

If you want something portable, consider using the appimage: https://github.com/iovisor/bpftrace/blob/master/INSTALL.md#appimage-install

**Reply by xiaoqing-wei** [reply]
Hi @danobi ,

just curious,
the binary from appimage looks linked with glibc 2.27, something like rhel8/centos8 level,
there are still many people using centos7 with a elrepo kernel (that is glibc 2.17 user space + latest upstream lts kernel like 5.4 /6.6 for now).

/data/bpftrace: /lib64/libm.so.6: version `GLIBC_2.27' not found (required by /data/bpftrace)
/data/bpftrace: /lib64/libc.so.6: version `GLIBC_2.27' not found (required by /data/bpftrace)
/data/bpftrace: /lib64/libc.so.6: version `GLIBC_2.22' not found (required by /data/bpftrace)



is there any plan to support lower glibc for the official static build ?

Thanks.

**Reply by danobi** [reply]
Hi @xiaoqing-wei , I'm not sure I follow. Are you seeing an error running appimage on older distro? If so, it'd be surprising to me - I expect that appimage ships its own libc

**Reply by xiaoqing-wei** [reply]
Hi @danobi ,
now I know it,

from 

the tagged official release builds are not ship with glibc, 
bpftrace binary sized 47M,
https://github.com/iovisor/bpftrace/releases/download/v0.19.1/bpftrace


and the bleeding edge builds, with its own glibc
bpftrace binary sized 160M
https://github.com/iovisor/bpftrace/actions/runs/7466099289/artifacts/1157743725

**Reply by viktormalik** [reply]
Hi @xiaoqing-wei, the reason is that AppImage support has been added after 0.19.1 release which still  has the old static binary which didn't include libc. The next 0.20 release should contain static binary with libc included (as bleeding edge builds do).

**Reply by xiaoqing-wei** [reply]
Hi @viktormalik , that's COOOL !
static binary with libc included is much more convenient.

**Reply by Engg-blogger** [reply]
@viktormalik So u r saying that from 0.20 release, the static binary (47 MB) in artifacts folder will be replaced with appimage (169 MB)? Or am i missing something?

**Reply by viktormalik** [reply]
Yes, exactly.

**Comment by Engg-blogger** [comment]
Ok.. makes sense. How can i build appimage for bpftrace myself? Could you guide me?

**Reply by danobi** [reply]
Please see https://github.com/iovisor/bpftrace/blob/master/docs/nix.md#build-bpftrace-as-a-statically-linked-binary

**Comment by Engg-blogger** [comment]
@danobi  So i tried downloading the appimage from the workflow page and then ran it, i got the error that i pasted earlier. But when i went back to an older commit and downloaded that version of appimage, it worked perfectly fine. https://github.com/iovisor/bpftrace/commit/6d80ae0ab1afb520fd1d42ba6b29d5a8c93d724a

So something broke the appimage for older distros in last one month.

**Reply by Engg-blogger** [reply]
@danobi @viktormalik This is the problematic commit : [280042efab50821e17918c1272762e0abd93c66f](https://dev.azure.com/msazure/One/_git/ASM-bpftrace/commit/280042efab50821e17918c1272762e0abd93c66f)
Appimages built after this commit are not working on older distros.

**Reply by viktormalik** [reply]
Thanks, this is very useful! I came to the same conclusion, let's continue the discussion in #2913.

**Comment by Engg-blogger** [comment]
@danobi One more question, could you let me know what distro and kernel version you are using to build the binary published here: https://github.com/iovisor/bpftrace/releases 
The reason why I am asking is that this published binary works on both old and latest kernel distros (from ubuntu 18 to 22). Whereas, when I built the binary on ubuntu 20 using the steps in install.MD, it did not work on ubuntu 22. Wanted to get some clarity on this.

**Reply by danobi** [reply]
The currently published release binaries were built with: https://github.com/iovisor/bpftrace/blob/3e368fa2dd6b1f9bc71a6fc0a0c1abe077b7d61c/.github/workflows/embedded.yml

**Reply by Engg-blogger** [reply]
Ok. I will take a look at it. One more question i have is if bpftrace is published on any PPA repository. The reason why I am asking is that when I install bpftrace from package manager in ubuntu 20.0 and below, it gives me version 0.9.0, which doesn't work. I was trying to see if bpftrace has any official PPA repository that i can add to ubuntu 20 to download the latest version.

**Reply by viktormalik** [reply]
No, we don't have a PPA repository. Instead, we provide the static builds to simplify getting the latest bpftrace on any distro.

**Reply by Engg-blogger** [reply]
@viktormalik If i want to create deb and rpm package for bpftrace. Whats the correct process to do that. Could you please guide?

**Reply by viktormalik** [reply]
We don't include instructions on building packages for distros in the repo. I'd suggest looking into the package in corresponding distros such as [Fedora](https://src.fedoraproject.org/rpms/bpftrace) or [Ubuntu](https://packages.ubuntu.com/jammy/bpftrace).
